terraform/transform_reference.go

```diff
@@ -106,18 +120,81 @@ func (t *ReferenceTransformer) Transform(g *Graph) error {
 	return nil
 }
 
+type depMap map[string]addrs.ConfigResource
+
+// addDep adds the vertex if it represents a resource in the
+// graph.
+func (m depMap) add(v dag.Vertex) {
+	// we're only concerned with resources which may have changes that
+	// need to be applied.
+	switch v := v.(type) {
+	case GraphNodeResourceInstance:
+		instAddr := v.ResourceInstanceAddr()
+		addr := instAddr.ContainingResource().Config()
+		m[addr.String()] = addr
+	case GraphNodeConfigResource:
+		addr := v.ResourceAddr()
+		m[addr.String()] = addr
+	}
+}
+
+// AttachDependsOnTransformer records all resources transitively referenced
+// through a configuration depends_on.
+type AttachDependsOnTransformer struct {
+}
+
+func (t AttachDependsOnTransformer) Transform(g *Graph) error {
+	// First we need to make a map of referenceable addresses to their vertices.
+	// This is very similar to what's done in ReferenceTransformer, but we keep
+	// implementation separate as they may need to change independently.
+	vertices := g.Vertices()
+	refMap := NewReferenceMap(vertices)
+
+	for _, v := range vertices {
+		depender, ok := v.(GraphNodeAttachDependsOn)
+		if !ok {
+			continue
+		}
+		selfAddr := depender.ResourceAddr()
+
+		// Only data need to attach depends_on, so they can determine if they
+		// are eligible to be read during plan.
+		if selfAddr.Resource.Mode != addrs.DataResourceMode {
```

Comment:
Would it _hurt_ to just do this unconditionally for every implementer of `GraphNodeAttachDependsOn`? It feels a little confusing to me that this is named as if it's a generic mechanism but then has this suprising restriction hidden in it, and my intuition from reading the rest of this PR is that other resource modes would just accept it and ignore it. :thinking:

Reply:
No, it wouldn't hurt. I did try to document in each location that these mechanisms only apply to data source, so I didn't want to attach them to managed resources and have something start using them from another code path (an occurrence which is unfortunately all too common)

Reply:
Fair enough! Perhaps we could ram that point home in a crude way by naming the interface `GraphNodeDataResourceAttachDependencies`, leaving no doubt at all that someone should think twice before using it for something other than a data resource, even if they don't read the comments.

(For similar reasons, I might also call this transformer `AttachDataResourceDependenciesTransformer`, so it's easy to see scanning the graph builder instantiation that this is something specifically for data resources.)